src/main/java/com/jfrog/ide/common/yarn/YarnTreeBuilder.java

```diff
@@ -93,6 +93,119 @@ private void addDepTreeNodes(Map<String, DepTreeNode> nodes, JsonNode jsonDep, D
+
+        if (splitPath == null) {
+            return null;
+        }
+
+        // packageFullName is guaranteed to be the last element in the path
+        if (!StringUtils.equals(splitPath[splitPath.length - 1], (StringUtils.substringBefore(packageFullName, ":")))) {
+            splitPath = Arrays.copyOf(splitPath, splitPath.length + 1);
+        }
+        splitPath[splitPath.length - 1] = packageFullName;
+        pathResult.addAll(Arrays.asList(splitPath));
+        return pathResult;
+    }
+
+    /**
+     * Extracts multiple dependency paths from a list of raw dependency Json strings returned from 'Yarn why' command.
+     *
+     * @param projectRootId      - The name of the project to display in the root of the impact tree.
+     * @param packageFullName    - The vulnerable dependency in <NAME>:<VERSION> format.
+     * @param rawDependencyPaths - The raw dependency Json strings returned from 'Yarn why' command.
+     * @return The extracted dependency paths as a list of dependencies starting from projectRootId till packageFullName.
+     */
+    List<List<String>> extractMultiplePaths(String projectRootId, String packageFullName, List<String> rawDependencyPaths) {
+        List<List<String>> paths = new ArrayList<>();
+        int limit = rawDependencyPaths.size() < ImpactTree.IMPACT_PATHS_LIMIT ? rawDependencyPaths.size() : 50;
+        for (int i = 0; i < limit; i++) {
+            List<String> path = extractSinglePath(projectRootId, packageFullName, rawDependencyPaths.get(i));
+            if (path != null) {
+                paths.add(path);
+            }
+        }
+        return paths;
+    }
+
+    /**
+     * Finds the dependency paths from the dependency to the root project, based on the supplied "yarn why" command output.
+     * A dependency path may appear as a part of a text or in a list of items.
+     * <p>
+     * Example 1 (Text):
+     * {"type":"info","data":"This module exists because \"jest-cli#istanbul-api#mkdirp\" depends on it."}
+     * <p>
+     * Example 2 (List):
+     * {"type":"list","data":{"type":"reasons","items":["Specified in \"dependencies\"","Hoisted from \"jest-cli#node-notifier#minimist\"","Hoisted from \"jest-cli#sane#minimist\""]}}
+     *
+     * @param logger          - The logger.
+     * @param projectRootId   - The name of the project to display in the root of the impact tree.
+     * @param packageName     - The package name (without version).
+     *                        Example: "minimist".
+     * @param packageVersions - The package versions.
+     * @return A map of package full name (<NAME>:<VERSION>) to a list of dependency paths.
+     * @throws IOException in case of I/O error returned from the running "yarn why" command in the yarnDriver.
+     */
+    public Map<String, List<List<String>>> findDependencyImpactPaths(Log logger, String projectRootId, String packageName, Set<String> packageVersions) throws IOException {
```

Comment:
The map can contains multiple keys in case the scanned project has the same package with multiple vulnerable versions. As the current design enforce Yarn why command to run once per package name, the map is required to attach the version-right impact graph to its version-corresponding vulnerable dependency node.